rfcs/text/design-tokens.md

```diff
@@ -0,0 +1,154 @@
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+    - [Design tokens via styled-components and CSS variables](#design-tokens-via-styled-components-and-css-variables)
+    - [Tag function for generating token](#tag-function-for-generating-token)
+- [Adoption strategy](#adoption-strategy)
+- [How we teach this](#how-we-teach-this)
+
+# Summary
+
+Implementation of the design tokens with theme support and backwards compatibility.
+
+# Basic example
+
+Design tokens will be kept in `@sage/design-tokens` library. It will be updated every time tokens will change. They will
+be imported from this library and used as the object.
+
+```js
+const Container = styled.div`
+  background: ${props.theme.sagedsColorsBase400};
+  color: ${props.theme.sagedsColorsTextWhite};
+`;
+```
+
+Then, themes would be handled by ThemeProvider.
+
+Themes would be created from existing ones by extending and renaming them, so that current *mint* would be named *sbiz*
+.
+
+```js
+import designTokens from "@sage/design-tokens/js/sbiz/es6-obj";
+import { mergeWithBase } from "../base";
+import configureTheme from "./mint-theme.config";
+
+export default {
+  ...mergeWithBase(configureTheme), // for backwards compability. Eventually should be removed.
+  ...designTokens,
+  name: "sbiz",
+};
+```
+
+# Motivation
+
+Design tokens are basic atomic part of design system. Its purpose is to improve handover process between designers and
+developers. This will result in more consistent ui and will help using colors more
+
+Design tokens are basically key-value pairs named and organized the same way regardless of the platform. This way,
+designers and developers can communicate using key names, and not specific values.
+
+Design tokens can contain values like lengths in different units, colors, font properties, transition definitions and
+others.
+
+Naming of the tokens is consistent across all platforms and provides predictable structure.
+
+To put it in simple words: When developer creates design for component, everything what you, as a developer, care about
+is which token has been used.
+
+# Detailed design
+
+This approach uses themes feature from `styled-components` since it is already implemented and uses javascript
+variables.
+
+At the moment, we have two levels of design tokens:
+
+- **Generic tokens** - kind of palette with all the values and scales.
+- **Theme tokens** - tokens for a theme, which contain references to generic tokens.
+
+So themes can be for example: *S Biz*, *M Biz*, *High Contrast*, *Dark Mode* and so on.
+
+Themes can overwrite not only colors, but also sizes. This way it is possible to create size themes, such as *Small*
+or *Medium* or *Large*
+
+In the end it all depends on designers, how they will build components in Figma, and which tokens they will use.
+
+# Drawbacks
+
+In case of CSS approach adaptation, there would be two theming systems available for carbon for the time of
+implementation.
+
+# Alternatives
```

Comment:
I would like to see 'Not using design tokens' listed as an alternative here.

i.e. what are design tokens bringing to this that we can't get from the existing styled component themes in Carbon.

Reply:
I'm open to suggestions for the final solution. CSS is more performant, however the we need to support having multiple versions of carbon on the page at the same time (duplicate CSS selectors?)

Reply:
> Aligning carbon with design system without design tokens wouldn't have much sense. One of puproses of design systems is to unify what designers do and what developers code, automate it and reduce possibilities of misunderstandings during the handover proceses. And that is what design tokens are for.

You are missing the point that Carbon is **the** implementation of the design system - therefore if it already supports theming - what is the use of design tokens bringing that gives us any additional benefit over what we have now.

Reply:
I didn't missed that. I see that carbon implements design system. However at the moment it is not implementing, which are the part of new Design System. That's why they will have to be implemented also in carbon.

> what is the use of design tokens bringing that gives us any additional benefit over what we have now.

Generally they are giving us consistency in all components across different platforms, and they are unifying language between designers, and developers in all technologies.

Design Tokens and theming are two, totally different terms and they are not connected directly, but theming functionality has impact on tokens naming.
- Theming is a feature, that allows users to change appearance of the library elements. As a functionality, it doesnt care about specific values.
- Design Tokens are sets of values that represents design decisions and are separated from the developers.

So in the end - themes can be extended with design tokens to keep this consistency.